server/src/main/java/org/elasticsearch/cluster/coordination/ClusterBootstrapService.java

```diff
@@ -21,214 +21,170 @@
             .anyMatch(s -> s.exists(settings));
     }
 
-    public void start() {
-        assert running == false;
-        running = true;
+    void onFoundPeersUpdated() {
+        final Set<DiscoveryNode> nodes = getDiscoveredNodes();
+        if (transportService.getLocalNode().isMasterNode() && initialMasterNodes.isEmpty() == false
+            && nodes.stream().noneMatch(Coordinator::isZen1Node) && checkWaitRequirements(nodes)) {
+
+            startBootstrap(nodes);
+        }
+    }
+
+    void scheduleUnconfiguredBootstrap() {
+        if (unconfiguredBootstrapTimeout == null) {
+            return;
+        }
 
         if (transportService.getLocalNode().isMasterNode() == false) {
             return;
         }
 
-        if (unconfiguredBootstrapTimeout != null) {
-            logger.info("no discovery configuration found, will perform best-effort cluster bootstrapping after [{}] " +
-                    "unless existing master is discovered", unconfiguredBootstrapTimeout);
-            final ThreadContext threadContext = transportService.getThreadPool().getThreadContext();
-            try (ThreadContext.StoredContext ignore = threadContext.stashContext()) {
-                threadContext.markAsSystemContext();
+        logger.info("no discovery configuration found, will perform best-effort cluster bootstrapping after [{}] " +
+            "unless existing master is discovered", unconfiguredBootstrapTimeout);
+
+        transportService.getThreadPool().scheduleUnlessShuttingDown(unconfiguredBootstrapTimeout, Names.GENERIC, new Runnable() {
+            @Override
+            public void run() {
+                final Set<DiscoveryNode> discoveredNodes = getDiscoveredNodes();
+                final List<DiscoveryNode> zen1Nodes = discoveredNodes.stream().filter(Coordinator::isZen1Node).collect(Collectors.toList());
+                if (zen1Nodes.isEmpty()) {
+                    logger.debug("performing best-effort cluster bootstrapping with {}", discoveredNodes);
+                    startBootstrap(discoveredNodes);
+                } else {
+                    logger.info("avoiding best-effort cluster bootstrapping due to discovery of pre-7.0 nodes {}", zen1Nodes);
+                }
+            }
+
+            @Override
+            public String toString() {
+                return "unconfigured-discovery delayed bootstrap";
+            }
+        });
+    }
+
+    private Set<DiscoveryNode> getDiscoveredNodes() {
+        return Stream.concat(Stream.of(transportService.getLocalNode()),
+            StreamSupport.stream(discoveredNodesSupplier.get().spliterator(), false)).collect(Collectors.toSet());
+    }
+
+    private void startBootstrap(Set<DiscoveryNode> discoveryNodes) {
+        if (bootstrappingPermitted.compareAndSet(true, false)) {
+            doBootstrap(new VotingConfiguration(discoveryNodes.stream().map(DiscoveryNode::getId).collect(Collectors.toSet())));
```

Comment:
also assert that all nodes here are master-eligible

Reply:
Sure, see c4555ac16a880be1c6f493dfd6590a89ceed0749.